VSR header validators; test parse_addresses

sentientwaffle opened the pull request.

sentientwaffle committed:
Unit-Tests: Unit tests for parse_addresses

sentientwaffle committed:
VSR: Add header validation

jorangreef commented on src/config.zig:
```diff
@@ -218,3 +218,10 @@ pub const clock_synchronization_window_min_ms = 2000;
+comptime {
+    const std = @import("std");
+
+    _ = std.net.Address.parseIp4(address, 0) catch unreachable;
```
Could we leave a comment here to explain why we run address/port through these?

jorangreef commented on src/vsr.zig:
```diff
@@ -223,10 +223,23 @@ pub const Header = packed struct {
         if (self.epoch != 0) return "epoch != 0";
         return switch (self.command) {
             .reserved => self.invalid_reserved(),
+            .ping => self.invalid_ping(),
```
Wow, these are a lot of validators when you see it all in place, nice work DJ!

Are these in the same order as we define them in `vsr.Header.Command`?

sentientwaffle replied:
Thanks!

> Are these in the same order as we define them in vsr.Header.Command?

Yes

jorangreef commented on src/vsr.zig:
```diff
@@ -590,6 +763,76 @@ pub fn parse_addresses(allocator: *std.mem.Allocator, raw: []const u8) ![]std.ne
 }
 
+test "parse_addresses" {
+    const test_successes = &[_]struct{
```
Great layout of these test success/failure vectors.

In the unit tests for StateMachine, we have `vectors` for this kind of test style. So let's call them `vectors_positive` and `vectors_negative` here, also to emphasize that we're thinking about positive/negative testing.

Then in the loops, we can capture them just as `vector`.

jorangreef commented on src/vsr.zig:
```diff
@@ -542,17 +712,20 @@ test "exponential_backoff_with_jitter" {
 /// This does require that the user specify the same order to all replicas.
 /// The caller owns the memory of the returned slice of addresses.
-/// TODO Unit tests.
 /// TODO Integrate into `src/cli.zig`.
```
Can we handle this CLI TODO as well?

If we can, then we know that what we're using in the CLI is the same as what we test here.

sentientwaffle replied:
The CLI is already using `vsr.parse_addresses` -- I'm not sure what the TODO is referring to.

https://github.com/coilhq/tigerbeetle/blob/main/src/cli.zig#L198-L199

jorangreef replied:
Ah then the TODO is stale, let's remove it.

jorangreef replied:
I believe I dealt with the TODO awhile back, but just didn't remove it.

jorangreef commented on src/vsr.zig:
```diff
@@ -336,6 +371,141 @@ pub const Header = packed struct {
 
+    fn invalid_reply(self: *const Header) ?[]const u8 {
+        assert(self.command == .reply);
+        if (self.client == 0) return "client == 0";
```
```suggestion
        // Initialization within `client.zig` asserts that client `id` is greater than zero:
        if (self.client == 0) return "client == 0";
```

jorangreef commented on src/vsr.zig:
```diff
@@ -336,6 +371,141 @@ pub const Header = packed struct {
+        assert(self.command == .reply);
+        if (self.client == 0) return "client == 0";
+        if (self.context != 0) return "context != 0";
+        if (self.op != self.commit) return "reserved: op != commit";
```
```suggestion
        if (self.op != self.commit) return "op != commit";
```

jorangreef commented on src/vsr.zig:
```diff
@@ -336,6 +371,141 @@ pub const Header = packed struct {
+        if (self.operation != .register) {
+            if (self.commit == 0) return "commit == 0";
+            if (self.request == 0) return "request == 0";
+        }
```
```suggestion
        if (self.operation == .register) {
            // In this context, the commit number is the newly registered session number.
            // The `0` commit number is reserved for cluster initialization.
            if (self.commit == 0) return "commit == 0";
            if (self.request != 0) return "request != 0";
        } else {
            if (self.commit == 0) return "commit == 0";
            if (self.request == 0) return "request == 0";
        }
```

jorangreef commented on src/vsr.zig:
```diff
@@ -336,6 +371,141 @@ pub const Header = packed struct {
+        assert(self.command == .commit);
+        if (self.parent != 0) return "parent != 0";
+        if (self.client != 0) return "client != 0";
+        if (self.context == 0) return "context == 0";
```
```suggestion
```

It's possible (albeit very rare 😁) that the cryptographic checksum context may be zero. I think we allow zero elsewhere, or was there another place where we rejected zero as a valid checksum/context?

sentientwaffle replied:
Is this a different situation?: https://github.com/coilhq/tigerbeetle/blob/dcaf309abb6e1650f2fb605c14efec3e44cf18a0/src/vsr.zig#L270

jorangreef replied:
Yes, it's a different situation. The `context` in this context (!) is simply storing the session number, which cannot be zero.

jorangreef commented on src/vsr.zig:
```diff
@@ -223,10 +223,23 @@ pub const Header = packed struct {
+            .reply => self.invalid_reply(),
+            .commit => self.invalid_commit(),
+            .start_view_change => self.invalid_start_view_change(),
+            .do_view_change, .start_view => self.invalid_do_view_change(),
```
Let's rather handle these explicitly using separate validators. There is a small difference in that the `start_view` message must have an offset of 0.

jorangreef commented on src/vsr.zig:
```diff
@@ -336,6 +371,141 @@ pub const Header = packed struct {
+        if (self.request != 0) return "request != 0";
+        if (self.offset != 0) return "offset != 0";
+        if (self.op == 0) return "op == 0";
+        if (self.operation != .reserved) return "operation != .reserved";
```
```suggestion
        // TODO implement validation once the message is actually used.
```

We may remove these recovery handlers, switching to persisting the view number to disk (with a few redundant copies on disk) before we ack a view transition externally as part of the view change protocol.

The rationale:

1. There's a correctness bug in the 2012 VSR paper for the recovery protocol, cf. https://syslab.cs.washington.edu/papers/diskless-tr16.pdf. However, the solution in there, if I understand correctly, seems to require unbounded vector clocks, which might be a little too complex. I am trying to get hold of the patched VSR implementation that the paper evaluated — it would be nice to see the impact of the fix on the actual recovery code.
2. The patched recovery protocol would still need a further protocol to break replicas out of recovery if all replicas were power-cycled at the same time, provided there is consensus on the cluster state.
3. Whereas, if we persist view numbers, then we're doing the 1988 Brian Oki flavor view change, so we can drop the round of `start_view_change` messages. The 2012 determinism plus the Brian Oki flavor is then pretty optimal: no dueling leaders, and less messages in the network, therefore less chance for network weather to impede the view change.

jorangreef replied:
Also just to clarify that if we do persist view numbers to disk, then there's no longer any need for the recovery protocol.

If we want to be able to handle all copies of on-disk metadata getting corrupted simultaneously (the PAR paper would simply panic in this case), then we can explore either the reconfiguration protocol or recovery protocol, but that's further down the road.

jorangreef commented on src/vsr.zig:
```diff
@@ -336,6 +371,141 @@ pub const Header = packed struct {
+        if (self.context != 0) return "context != 0";
+        if (self.request != 0) return "request != 0";
+        if (self.offset != 0) return "offset != 0";
+        if (self.op == 0) return "op == 0";
```
```suggestion
        // TODO Add local recovery mechanism for repairing the cluster initialization "zero" op.
        if (self.op == 0) return "op == 0";
        if (self.commit > self.op) return "op_min > op_max";
```

jorangreef commented on src/vsr.zig:
```diff
@@ -542,17 +712,20 @@ test "exponential_backoff_with_jitter" {
     errdefer allocator.free(addresses);
 
     var index: usize = 0;
     var comma_iterator = std.mem.split(raw, ",");
```
```suggestion
    var comma_iterator = std.mem.split(u8, raw, ",");
```

We can land this after the `zig-0.9.0` branch lands on main, so we can already target Zig 0.9.0.

sentientwaffle replied:
:+1: I'll rebase once `zig-0.9.0` is merged.

jorangreef commented on src/vsr.zig:
```diff
@@ -590,6 +763,76 @@ pub fn parse_addresses(allocator: *std.mem.Allocator, raw: []const u8) ![]std.ne
+    }{
+        .{
+            // Test the minimum/maximum port.
+            .raw = "1.2.3.4:0,2.3.4.5:789,3.4.5.6:65535",
```
Great testing of the 0 and max 16-bit boundaries here.

Let's do the same for the IPv4 components as well, have one address of "0.0.0.0", another of "255.255.255.255", and one where the components are all distinct like we have already "1.2.3.4"?

jorangreef reviewed (changes requested):
Great work @sentientwaffle! Nice to have these validators and tests in place.

sentientwaffle committed:
VSR: code review: validation fixes/cleanup

- Split `invalid_do_view_change` and `invalid_start_view`.
- `commit` can be `0`.
- `recovery_response` is unused for now.

jorangreef commented:
@sentientwaffle ready to land when you think it's ready. Please merge when you think it's ready.
